driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -201,19 +201,32 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> ou
             // version first bit is the "direction" of the frame (request or response)
             version = version & 0x7F;
 
-            DecoderForStreamIdSize decoder = (version >= 3) ? decoderV3 : decoderV1;
+            DecoderForStreamIdSize decoder = decoderForStreamIdSize.get();
+            if (decoder == null) {
+                decoder = new DecoderForStreamIdSize(version, version >= 3 ? 2 : 1);
+                if (!decoderForStreamIdSize.compareAndSet(null, decoder)) {
+                    decoder = decoderForStreamIdSize.get();
+                }
+            }
+
+            // the version in the message should match the version this Decoder is configured for.
+            assert decoder.protocolVersion == version;
+
             Object frame = decoder.decode(ctx, buffer);
             if (frame != null)
                 out.add(frame);
         }
 
         static class DecoderForStreamIdSize extends LengthFieldBasedFrameDecoder {
-            private static final int MAX_FRAME_LENGTH = 256 * 1024 * 1024; // 256 MB
+            // The maximum response frame length allowed.  Note that C* does not currently restrict the length of its responses (CASSANDRA-12630).
+            private static final int MAX_FRAME_LENGTH = SystemProperties.getInt("com.datastax.driver.NATIVE_TRANSPORT_MAX_FRAME_SIZE_IN_MB", 256) * 1024 * 1024; // 256 MB
```

Comment:
This could also go into `ProtocolOptions`, however the annoying thing with that class is that every new field requires a new `Cluster.Builder` method (there is no `withProtocolOptions()`). Changing the max frame size is an edge case, so I guess this can stay as a system property.